### ISSUE TITLE
cmd/distpack: release archives don't include directory members

### ISSUE THREAD

**Issue by jeffreytolar** [golang/go#61862]
### What version of Go are you using (`go version`)?

<pre>
# /usr/local/go/bin/go version
go version go1.21.0 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
# /usr/local/go/bin/go env
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/root/.cache/go-build'
GOENV='/root/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/root/go/pkg/mod'
GONOPROXY='go.ourcompany.com'
GONOSUMDB='go.ourcompany.com'
GOOS='linux'
GOPATH='/root/go'
GOPRIVATE='go.ourcompany.com'
GOPROXY='https://our-internal-goproxy.ourcompany.com,https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.0'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2916500537=/tmp/go-build -gno-record-gcc-switches'
</pre></details>

### What did you do?

```
# curl -L --remote-name-all https://go.dev/dl/go{1.20,1.21.0}.linux-amd64.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    73  100    73    0     0    327      0 --:--:-- --:--:-- --:--:--   328
100 95.2M  100 95.2M    0     0  73.2M      0  0:00:01  0:00:01 --:--:--  107M
100    75  100    75    0     0   2027      0 --:--:-- --:--:-- --:--:--  2027
100 63.3M  100 63.3M    0     0   117M      0 --:--:-- --:--:-- --:--:--  144M


# sha256sum go*.tar.gz
5a9ebcc65c1cce56e0d2dc616aff4c4cedcfbda8cc6f0288cc08cda3b18dcbf1  go1.20.linux-amd64.tar.gz
d0398903a16ba2232b389fb31032ddf57cac34efda306a0eebac34f0965a0742  go1.21.0.linux-amd64.tar.gz


# tar tzvf go1.20.linux-amd64.tar.gz | grep ' go/[^/]*/$'
drwxr-xr-x root/root         0 2023-01-30 21:48 go/api/
drwxr-xr-x root/root         0 2023-01-30 21:49 go/bin/
drwxr-xr-x root/root         0 2023-01-30 21:48 go/doc/
drwxr-xr-x root/root         0 2023-01-30 21:48 go/lib/
drwxr-xr-x root/root         0 2023-01-30 21:48 go/misc/
drwxr-xr-x root/root         0 2023-01-30 21:49 go/pkg/
drwxr-xr-x root/root         0 2023-01-30 21:48 go/src/
drwxr-xr-x root/root         0 2023-01-30 21:48 go/test/


# tar tzvf go1.21.0.linux-amd64.tar.gz | grep ' go/[^/]*/$'
(empty)
# tar tzvf go1.21.0.linux-amd64.tar.gz | grep -c '/$'
0
```

### What did you expect to see?

Directory entries in the release tarball

### What did you see instead?

No directory entries in the release tarball

- - -

I'm not entirely sure if this is a bug or if it's an intentional change; we noticed it on hosts with a restricted `umask`, since the extracted directories didn't have world-read/execute permissions (unlike past releases).

I searched the issue tracker for "distpack", but didn't see anything that looked relevant.

This is presumably a consequence of 1635205a72d26721af54f01ccbab8e0b51ded3a9 (switching to distpack for 1.21) / #24904?

**Comment by dmitshur** [comment]
The documentation of [`distpack.Archive`](https://cs.opensource.google/go/go/+/master:src/cmd/distpack/archive.go;l=19;drc=3afa5daabd6245be0c03b6e666f93835e32d7f7c) suggests it was an intentional decision at the time:

> Directories are implied by the files and not explicitly listed.

As does the second bullet point in #61513.

CC @rsc, @heschi.

**Comment by heschi** [comment]
Yeah, it was an intentional change. Because of the reproducibility push, changing back would be a bit troublesome. Offhand, it seems like honoring the host's umask is reasonable?

**Comment by jeffreytolar** [comment]
Ok; it results in some odd behavior, however - GNU tar defaults to `--preserve-permissions`, which means it'll use the file modes in the tarball. With a umask of `027`, this results in only the directories having the restrictive permissions.

E.g., on rhel8:
```
bash-4.4# curl -L --remote-name-all https://go.dev/dl/go{1.20,1.21.0}.linux-amd64.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    73  100    73    0     0    304      0 --:--:-- --:--:-- --:--:--   305
100 95.2M  100 95.2M    0     0  72.8M      0  0:00:01  0:00:01 --:--:--  110M
100    75  100    75    0     0   1923      0 --:--:-- --:--:-- --:--:--  1923
100 63.3M  100 63.3M    0     0  91.4M      0 --:--:-- --:--:-- --:--:--  142M

bash-4.4# umask 027
bash-4.4# mkdir /usr/local/go{1.20,1.21.0}
bash-4.4# tar xzf go1.20.linux-amd64.tar.gz -C /usr/local/go1.20/
bash-4.4# tar xzf go1.21.0.linux-amd64.tar.gz -C /usr/local/go1.21.0/


bash-4.4# ls -ld /usr/local/go{1.20,1.21.0}/go
drwxr-xr-x. 10 root root 4096 Jan 30  2023 /usr/local/go1.20/go
drwxr-x---. 10 root root 4096 Aug  8 22:52 /usr/local/go1.21.0/go

bash-4.4# ls -ld /usr/local/go{1.20,1.21.0}/go/bin/go
-rwxr-xr-x. 1 root root 15563590 Jan 30  2023 /usr/local/go1.20/go/bin/go
-rwxr-xr-x. 1 root root 12403672 Aug  4 20:14 /usr/local/go1.21.0/go/bin/go
```

(Looks like rhel8's `bsdtar` has the same behavior)

**Comment by rsc** [comment]
We will fix this for Go 1.21.1. For supply chain reasons we can't reissue the 1.21.0 files.

**Comment by heschi** [comment]
Dry-run build looks good to me, including the signed Darwin archives.
